raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -816,26 +815,63 @@ public void testInvalidFetchRequest() throws Exception {
+            fetchQuorumRecordsRequest(epoch, otherNodeId, -5L, 0, 0), time.milliseconds()));
         client.poll();
         assertFailedFetchQuorumRecordsResponse(Errors.INVALID_REQUEST);
 
         channel.mockReceive(new RaftRequest.Inbound(channel.newCorrelationId(),
-            fetchQuorumRecordsRequest(epoch, otherNodeId, 0L, -1), time.milliseconds()));
+            fetchQuorumRecordsRequest(epoch, otherNodeId, 0L, -1, 0), time.milliseconds()));
         client.poll();
         assertFailedFetchQuorumRecordsResponse(Errors.INVALID_REQUEST);
 
         channel.mockReceive(new RaftRequest.Inbound(channel.newCorrelationId(),
-            fetchQuorumRecordsRequest(epoch, otherNodeId, 0L, epoch + 1), time.milliseconds()));
+            fetchQuorumRecordsRequest(epoch, otherNodeId, 0L, epoch + 1, 0), time.milliseconds()));
         client.poll();
         assertFailedFetchQuorumRecordsResponse(Errors.INVALID_REQUEST);
 
         channel.mockReceive(new RaftRequest.Inbound(channel.newCorrelationId(),
-            fetchQuorumRecordsRequest(epoch + 1, otherNodeId, 0L, 0), time.milliseconds()));
+            fetchQuorumRecordsRequest(epoch + 1, otherNodeId, 0L, 0, 0), time.milliseconds()));
         client.poll();
         assertFailedFetchQuorumRecordsResponse(Errors.INVALID_REQUEST);
     }
 
+    @Test
+    public void testPurgatoryFetchTimeout() throws Exception {
+        int otherNodeId = 1;
+        int epoch = 5;
+
+        quorumStateStore.writeElectionState(ElectionState.withElectedLeader(epoch, localId));
+        Set<Integer> voters = Utils.mkSet(localId, otherNodeId);
+        KafkaRaftClient client = buildClient(voters);
+        assertEquals(ElectionState.withElectedLeader(epoch, localId), quorumStateStore.readElectionState());
+
+        // Handle FindQuorum
+        pollUntilSend(client);
+        int findQuorumCorrelationId = assertSentFindQuorumRequest();
+        channel.mockReceive(new RaftResponse.Inbound(findQuorumCorrelationId,
+            findQuorumResponse(OptionalInt.empty(), 1, voters), -1));
+
+        // Handle BeginEpoch
+        pollUntilSend(client);
+        int beginEpochCorrelationId = assertBeginQuorumEpochRequest(epoch);
+        channel.mockReceive(new RaftResponse.Inbound(beginEpochCorrelationId,
+            beginEpochResponse(epoch, localId), otherNodeId));
+        client.poll();
+
+        // Follower sends a fetch which cannot be satisfied immediately
+        int maxWaitTimeMs = 500;
+        FetchQuorumRecordsRequestData fetchRequest = fetchQuorumRecordsRequest(
+            epoch, otherNodeId, 1L, epoch, maxWaitTimeMs);
+        channel.mockReceive(new RaftRequest.Inbound(channel.newCorrelationId(), fetchRequest, time.milliseconds()));
+        client.poll();
+        assertEquals(0, channel.drainSendQueue().size());
+
+        // After expiration of the max wait time, the fetch returns
+        time.sleep(maxWaitTimeMs);
```

Comment:
What if we sleep longer here?

Reply:
I guess my assumption is that waiting longer would result in the same behavior. Do you think it is worth testing this separately?